From 7d8ca05eec49748d8c1ea9fcf968d4f01e476dfd Mon Sep 17 00:00:00 2001
From: "Guoniu.Zhou" <guoniu.zhou@nxp.com>
Date: Mon, 22 Oct 2018 15:48:00 +0800
Subject: [PATCH 5002/5242] MMFMWK-8296: ISI: return error code when user try
 to do upscale

commit  d299073c052e9306667100f42c0c3f5629f5e1ac from
https://source.codeaurora.org/external/imx/linux-imx.git

Return error code when user try to do upscale since isi don't support.

Signed-off-by: Guoniu.Zhou <guoniu.zhou@nxp.com>
(cherry picked from commit 59949e29dcd996c50cae43bc5ff3f89907761c1b)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/imx8/mxc-isi-cap.c |    7 +++++++
 drivers/media/platform/imx8/mxc-isi-m2m.c |    9 +++++++++
 2 files changed, 16 insertions(+)

diff --git a/drivers/media/platform/imx8/mxc-isi-cap.c b/drivers/media/platform/imx8/mxc-isi-cap.c
index 7d9ae38..564fdef 100644
--- a/drivers/media/platform/imx8/mxc-isi-cap.c
+++ b/drivers/media/platform/imx8/mxc-isi-cap.c
@@ -873,6 +873,7 @@ static int mxc_isi_cap_s_fmt_mplane(struct file *file, void *priv,
 	struct mxc_isi_dev *mxc_isi = video_drvdata(file);
 	struct v4l2_pix_format_mplane *pix = &f->fmt.pix_mp;
 	struct mxc_isi_frame *dst_f = &mxc_isi->isi_cap.dst_f;
+	struct mxc_isi_frame *src_f = &mxc_isi->isi_cap.src_f;
 	struct mxc_isi_fmt *fmt;
 	int bpl;
 	int i;
@@ -947,6 +948,12 @@ static int mxc_isi_cap_s_fmt_mplane(struct file *file, void *priv,
 
 	mxc_isi_source_fmt_init(mxc_isi);
 
+	if ((dst_f->width  > src_f->width) ||
+		(dst_f->height > src_f->height)) {
+		dev_err(&mxc_isi->pdev->dev, "%s: Not support upscale\n", __func__);
+		return -EINVAL;
+	}
+
 	mxc_isi_channel_init(mxc_isi);
 	/* configure mxc isi channel */
 	mxc_isi_channel_config(mxc_isi);
diff --git a/drivers/media/platform/imx8/mxc-isi-m2m.c b/drivers/media/platform/imx8/mxc-isi-m2m.c
index 47f1def..65c7df0 100644
--- a/drivers/media/platform/imx8/mxc-isi-m2m.c
+++ b/drivers/media/platform/imx8/mxc-isi-m2m.c
@@ -800,8 +800,17 @@ static int mxc_isi_m2m_streamon(struct file *file, void *priv,
 			     enum v4l2_buf_type type)
 {
 	struct mxc_isi_dev *mxc_isi = video_drvdata(file);
+	struct mxc_isi_frame *src_f, *dst_f;
 	int ret;
 
+	src_f = &mxc_isi->m2m.src_f;
+	dst_f = &mxc_isi->m2m.dst_f;
+	if ((dst_f->width  > src_f->width) ||
+		(dst_f->height > src_f->height)) {
+		dev_err(&mxc_isi->pdev->dev, "%s Not support upscale\n", __func__);
+		return -EINVAL;
+	}
+
 	if (type == V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE) {
 		mxc_isi->m2m.frame_count = 0;
 		mxc_isi_m2m_channel_config(mxc_isi);
-- 
1.7.9.5

